src/probability/martingale/borel_cantelli.lean

```diff
@@ -0,0 +1,400 @@
+  ℱ.mono (finset.mem_range.1 hk) _ $ hs _
+
+lemma martingale_part_process_ae_eq (ℱ : filtration ℕ m0) (s : ℕ → set Ω) (n : ℕ) :
+  martingale_part ℱ μ (process s) n =ᵐ[μ]
+  ∑ k in finset.range n, ((s (k + 1)).indicator 1 - μ[(s (k + 1)).indicator 1 | ℱ k]) :=
+begin
+  simp only [martingale_part_eq_sum, process_zero, zero_add],
+  refine eventually_eq_sum (λ k hk, _),
+  simp only [process, finset.sum_range_succ_sub_sum],
+end
+
+lemma martingale_part_process_ae_eq' (ℱ : filtration ℕ m0) (s : ℕ → set Ω) :
+  ∀ᵐ ω ∂μ, ∀ n, martingale_part ℱ μ (process s) n ω =
+  ∑ k in finset.range n, ((s (k + 1)).indicator 1 - μ[(s (k + 1)).indicator 1 | ℱ k]) ω :=
+begin
+  rw ae_all_iff,
+  intro n,
+  convert martingale_part_process_ae_eq ℱ s n,
+  ext ω,
+  rw finset.sum_apply,
+end
+
+variables [is_finite_measure μ]
+
+lemma integrable_process (hs : ∀ n, measurable_set[ℱ n] (s n)) (n : ℕ) :
+  integrable (process s n) μ :=
+integrable_finset_sum' _ $ λ k hk, integrable_on.indicator (integrable_const 1) $ ℱ.le _ _ $ hs _
+
+lemma martingale_part_process_diff_le (hs : ∀ n, measurable_set[ℱ n] (s n)) (n : ℕ) :
+  ∀ᵐ ω ∂μ, |martingale_part ℱ μ (process s) (n + 1) ω - martingale_part ℱ μ (process s) n ω| ≤ 1 :=
+begin
+  have h₁ : μ[(s (n + 1)).indicator 1|ℱ n] ≤ᵐ[μ] 1,
+  { change _ ≤ᵐ[μ] (λ ω, 1 : Ω → ℝ),
+    rw ← @condexp_const _ _ _ _ _ _ _ μ (ℱ.le n) (1 : ℝ),
+    refine condexp_mono ((integrable_indicator_iff (ℱ.le _ _ (hs $ _))).2
+      (integrable_const 1).integrable_on) (integrable_const 1)
+      (eventually_of_forall $ λ ω, set.indicator_le_self' (λ _ _, zero_le_one) ω) },
+  have h₂ : (0 : Ω → ℝ) ≤ᵐ[μ] μ[(s (n + 1)).indicator 1|ℱ n],
+    from condexp_nonneg (eventually_of_forall $ λ ω, set.indicator_nonneg (λ _ _, zero_le_one) _),
+  filter_upwards [h₁, h₂, martingale_part_process_ae_eq ℱ s (n + 1),
+    martingale_part_process_ae_eq ℱ s n] with ω hω₁ hω₂ hω₃ hω₄,
+  simp_rw [hω₃, hω₄, finset.sum_apply, finset.sum_range_succ_sub_sum],
+  rw [abs_le, neg_le, pi.sub_apply, neg_sub, tsub_le_iff_right, tsub_le_iff_right,
+    add_comm (1 : ℝ), add_comm (1 : ℝ)],
+  exact ⟨le_add_of_nonneg_of_le (set.indicator_nonneg (λ _ _, zero_le_one) _) hω₁,
+    le_add_of_nonneg_of_le hω₂ (set.indicator_le' (λ _ _, le_rfl) (λ _ _, zero_le_one) ω)⟩,
+end
+
+lemma martingale_part_process_diff_le' (hs : ∀ n, measurable_set[ℱ n] (s n)) :
+  ∀ᵐ ω ∂μ, ∀ n,
+    |martingale_part ℱ μ (process s) (n + 1) ω - martingale_part ℱ μ (process s) n ω| ≤ (1 : ℝ≥0) :=
+begin
+  rw [ae_all_iff, nonneg.coe_one],
+  exact martingale_part_process_diff_le hs ,
+end
+
+end borel_cantelli
+
+open borel_cantelli
+
+lemma tendsto_sum_indicator_at_top_iff
+  [is_finite_measure μ] {s : ℕ → set Ω} (hs : ∀ n, measurable_set[ℱ n] (s n)) :
+  ∀ᵐ ω ∂μ,
+    tendsto (λ n, ∑ k in finset.range n, (s (k + 1)).indicator (1 : Ω → ℝ) ω) at_top at_top ↔
+    tendsto (λ n, ∑ k in finset.range n, μ[(s (k + 1)).indicator (1 : Ω → ℝ) | ℱ k] ω)
+      at_top at_top :=
```

Comment:
Is this lemma specific to that process, or could it be extended to any adapted process and its predictable part under reasonable conditions (bounded increments perhaps)?

Reply:
I've generalized it to 
```lean
lemma tendsto_sum_indicator_at_top_iff
  [is_finite_measure μ] (hfmono : ∀ᵐ ω ∂μ, ∀ n, f n ω ≤ f (n + 1) ω)
  (hf : adapted ℱ f) (hint : ∀ n, integrable (f n) μ)
  (hbdd : ∀ᵐ ω ∂μ, ∀ n, |f (n + 1) ω - f n ω| ≤ R) :
  ∀ᵐ ω ∂μ, tendsto (λ n, f n ω) at_top at_top ↔
    tendsto (λ n, predictable_part ℱ μ f n ω) at_top at_top
```